src/Compilers/CSharp/Portable/Symbols/Source/SourceMethodSymbolWithAttributes.cs

```diff
@@ -808,6 +812,88 @@ private void DecodeModuleInitializerAttribute(DecodeWellKnownAttributeArguments<
+            }
+
+            if (!arguments.Attribute.CommonNamedArguments.IsDefaultOrEmpty)
+            {
+                foreach (var (key, value) in arguments.Attribute.CommonNamedArguments)
+                {
+                    if (key != "CallConvs"
+                        || value.Kind != TypedConstantKind.Array
+                        || value.Values.Any(v => v.Kind != TypedConstantKind.Type))
+                    {
+                        continue;
+                    }
+
+                    foreach (var callConvTypedConstant in value.Values)
+                    {
+                        Debug.Assert(callConvTypedConstant.Kind == TypedConstantKind.Type);
+                        if (!(callConvTypedConstant.ValueInternal is NamedTypeSymbol callConvType)
+                            || !FunctionPointerTypeSymbol.IsCallingConventionModifier(callConvType))
+                        {
+                            // `{0}` is not a valid calling convention type for 'UnmanagedCallersOnly'.
+                            arguments.Diagnostics.Add(ErrorCode.ERR_InvalidUnmanagedCallersOnlyCallConv, arguments.AttributeSyntaxOpt.Location, callConvTypedConstant.ValueInternal ?? "null");
+                        }
+                    }
+                }
+            }
+
+            var returnTypeSyntax = SyntaxNode switch
+            {
+                MethodDeclarationSyntax m => m.ReturnType,
+                LocalFunctionStatementSyntax l => l.ReturnType,
+                _ => null
+            };
+
+            if (returnTypeSyntax is null)
+            {
+                return;
+            }
+
+            checkAndReportManagedTypes(ReturnType, returnTypeSyntax, isParam: false, arguments.Diagnostics);
+            foreach (var param in Parameters)
+            {
+                checkAndReportManagedTypes(param.Type, param.GetNonNullSyntaxNode(), isParam: true, arguments.Diagnostics);
+            }
+
+            static void checkAndReportManagedTypes(TypeSymbol type, SyntaxNode syntax, bool isParam, DiagnosticBag diagnostics)
+            {
+                // use-site diagnostics will be reported at actual parameter declaration site, we're only interested
+                // in reporting managed types being used
+                switch (type.ManagedKindNoUseSiteDiagnostics)
+                {
+                    case ManagedKind.Unmanaged:
+                        return;
+
+                    case ManagedKind.UnmanagedWithGenerics:
+                        // If the type has generics, do a check on the generic types. If they're constrained to be unmanaged,
+                        // then this is valid.
+                        Debug.Assert(type is NamedTypeSymbol { IsGenericType: true });
+                        if (((NamedTypeSymbol)type).TypeParameters.Any(parameter => parameter.ManagedKindNoUseSiteDiagnostics != ManagedKind.Unmanaged))
```

Comment:
>TypeParameters [](start = 52, length = 14)

Should this be `TypeArguments`?

Reply:
Clarified with @elinor-fung on what the correct requirements here are. After discussion, they are: All parameter and return types must be truly unmanaged (ie, not UnmanagedWithGenerics), and the method itself cannot have generic type parameters, even if those parameters are constrained to unmanaged themselves.

---
In reply to: [476649712](https://github.com/dotnet/roslyn/pull/47008#discussion_r476649712) [](ancestors = 476649712)